File: scala/parsers/src/test/scala/jp/co/teruuu/parser/common/OrParserTest.scala
package jp.co.teruuu.parser.common

import org.scalatest.FlatSpec;

class OrParserTest extends FlatSpec {

  "parse" should "successt" in {
    val parser = Parser.string("Hello").or(Parser.string("World"))

    val result1 = parser.parse("HelloWorld")
    assert(result1.isInstanceOf[ParseOk[_]])
    assert(result1.asInstanceOf[ParseOk[String]].value.equals("Hello"))

    val result2 = parser.parse("WorldHello")
    assert(result2.isInstanceOf[ParseOk[_]])
    assert(result2.asInstanceOf[ParseOk[String]].value.equals("World"))

    val result3 = parser.parse("helloWorld")
    assert(result3.isInstanceOf[ParseNg[_]])
  }

}

File: java/parsers_java8/src/test/java/jp/co/teruuu/parser/common/OrParserTest.java
package jp.co.teruuu.parser.common;

import jp.co.teruuu.parser.common.ParseResult;
import jp.co.teruuu.parser.common.Parser;
import org.junit.Test;

public class OrParserTest {

    @Test
    public void test() {
        Parser<String> orParser = Parser.string("Hello").or(Parser.string("World"));
        ParseResult result1 = orParser.parse("Hello World");
        assert(result1 instanceof ParseResult.Success);
        assert(((ParseResult.Success) result1).value.equals("Hello"));

        ParseResult result2 = orParser.parse("World Hello");
        assert(result2 instanceof ParseResult.Success);
        assert(((ParseResult.Success) result2).value.equals("World"));

        ParseResult result3 = orParser.parse("world Hello");
        assert(result3 instanceof ParseResult.Failure);
    }
}
